File: test/custom_backend/test_custom_backend.py
# Owner(s): ["module: unknown"]

import os
import tempfile
import torch

from backend import Model, to_custom_backend, get_custom_backend_library_path
from torch.testing._internal.common_utils import TestCase, run_tests


class TestCustomBackend(TestCase):
    def setUp(self):
        # Load the library containing the custom backend.
        self.library_path = get_custom_backend_library_path()
        torch.ops.load_library(self.library_path)
        # Create an instance of the test Module and lower it for
        # the custom backend.
        self.model = to_custom_backend(torch.jit.script(Model()))

    def test_execute(self):
        """
        Test execution using the custom backend.
        """
        a = torch.randn(4)
        b = torch.randn(4)
        # The custom backend is hardcoded to compute f(a, b) = (a + b, a - b).
        expected = (a + b, a - b)
        out = self.model(a, b)
        self.assertTrue(expected[0].allclose(out[0]))
        self.assertTrue(expected[1].allclose(out[1]))

    def test_save_load(self):
        """
        Test that a lowered module can be executed correctly
        after saving and loading.
        """
        # Test execution before saving and loading to make sure
        # the lowered module works in the first place.
        self.test_execute()

        # Save and load.
        f = tempfile.NamedTemporaryFile(delete=False)
        try:
            f.close()
            torch.jit.save(self.model, f.name)
            loaded = torch.jit.load(f.name)
        finally:
            os.unlink(f.name)
        self.model = loaded

        # Test execution again.
        self.test_execute()


if __name__ == "__main__":
    run_tests()

File: test/custom_backend/test_custom_backend.cpp
#include <torch/cuda.h>
#include <torch/script.h>

#include <string>

#include "custom_backend.h"

// Load a module lowered for the custom backend from \p path and test that
// it can be executed and produces correct results.
void load_serialized_lowered_module_and_execute(const std::string& path) {
  torch::jit::Module module = torch::jit::load(path);
  // The custom backend is hardcoded to compute f(a, b) = (a + b, a - b).
  auto tensor = torch::ones(5);
  std::vector<torch::jit::IValue> inputs{tensor, tensor};
  auto output = module.forward(inputs);
  AT_ASSERT(output.isTuple());
  auto output_elements = output.toTuple()->elements();
  for (auto& e : output_elements) {
    AT_ASSERT(e.isTensor());
  }
  AT_ASSERT(output_elements.size(), 2);
  AT_ASSERT(output_elements[0].toTensor().allclose(tensor + tensor));
  AT_ASSERT(output_elements[1].toTensor().allclose(tensor - tensor));
}

int main(int argc, const char* argv[]) {
  if (argc != 2) {
    std::cerr
        << "usage: test_custom_backend <path-to-exported-script-module>\n";
    return -1;
  }
  const std::string path_to_exported_script_module = argv[1];

  std::cout << "Testing " << torch::custom_backend::getBackendName() << "\n";
  load_serialized_lowered_module_and_execute(path_to_exported_script_module);

  std::cout << "OK\n";
  return 0;
}
